File: test-runtime/src/main/java/dev/vibris/testruntime/FakeVibrisServerMain.java
package dev.vibris.testruntime;

import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.file.Path;

public final class FakeVibrisServerMain {
    private FakeVibrisServerMain() {
    }

    public static void main(String[] arguments) throws Exception {
        Options options = Options.parse(arguments);
        FakeVibrisServer server = FakeVibrisServer.start(
            options.port(), options.pendingRoot(), options.artifactRoot());
        Thread shutdownHook = new Thread(() -> close(server), "Vibris Test Runtime Shutdown");
        Runtime.getRuntime().addShutdownHook(shutdownHook);
        try {
            if (options.probeControlStdin()) {
                awaitProbeShutdown();
                server.close();
                ProbeJson.write(server.phaseThreeProbe().snapshot(),
                    server.phaseThreeProbe().maxConcurrentRuntimeOperations(), System.out);
            } else {
                server.awaitTermination();
            }
        } finally {
            try {
                Runtime.getRuntime().removeShutdownHook(shutdownHook);
            } catch (IllegalStateException ignored) {
            }
            server.close();
        }
    }

    private static void awaitProbeShutdown() throws Exception {
        BufferedReader input = new BufferedReader(new InputStreamReader(System.in));
        String line;
        while ((line = input.readLine()) != null) {
            if (!line.isEmpty() && line.charAt(0) == '\uFEFF') line = line.substring(1);
            if (line.equals("shutdown")) return;
            if (!line.isBlank()) throw new IllegalArgumentException("Unknown probe control command: " + line);
        }
    }

    private static void close(FakeVibrisServer server) {
        try {
            server.close();
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
        }
    }

    private record Options(int port, Path workRoot, Path pendingRoot, Path artifactRoot, boolean probeControlStdin) {
        private static Options parse(String[] arguments) {
            Integer port = null;
            Path workRoot = null;
            Path pendingRoot = null;
            Path artifactRoot = null;
            boolean probeControlStdin = false;
            for (int index = 0; index < arguments.length; index++) {
                switch (arguments[index]) {
                    case "--port" -> port = Integer.parseInt(requireValue(arguments, ++index, "--port"));
                    case "--work-root" -> workRoot = Path.of(requireValue(arguments, ++index, "--work-root"));
                    case "--pending-root" -> pendingRoot = Path.of(requireValue(arguments, ++index, "--pending-root"));
                    case "--artifact-root" ->
                            artifactRoot = Path.of(requireValue(arguments, ++index, "--artifact-root"));
                    case "--probe-control-stdin" -> probeControlStdin = true;
                    default -> throw new IllegalArgumentException("Unknown argument: " + arguments[index]);
                }
            }
            if (port == null) throw new IllegalArgumentException("Missing required argument: --port");
            if (workRoot == null) throw new IllegalArgumentException("Missing required argument: --work-root");
            Path root = workRoot.toAbsolutePath().normalize();
            Path resolvedPending = pendingRoot == null ? root.resolve("pending-shaders") : pendingRoot;
            Path resolvedArtifacts = artifactRoot == null ? root.resolve("artifacts") : artifactRoot;
            return new Options(port, root, resolvedPending, resolvedArtifacts, probeControlStdin);
        }

        private static String requireValue(String[] arguments, int index, String option) {
            if (index >= arguments.length) throw new IllegalArgumentException("Missing value for " + option);
            return arguments[index];
        }
    }
}

File: test-runtime/src/main/kotlin/dev/vibris/testruntime/FakeVibrisServerMain.kt
package dev.vibris.testruntime

import dev.vibris.core.VibrisBootstrap
import java.io.BufferedReader
import java.io.InputStreamReader
import java.nio.file.Path

object FakeVibrisServerMain {
    @JvmStatic
    @Throws(Exception::class)
    fun main(arguments: Array<String>) {
        val options = Options.parse(arguments)
        val shaderpackRoot = options.shaderpackRoot
        if (shaderpackRoot != null) {
            runBootstrap(options, shaderpackRoot)
            return
        }
        val server = FakeVibrisServer.start(options.port, options.pendingRoot, options.artifactRoot)
        val shutdownHook = Thread(
            { close(server) },
            "Vibris Test Runtime Shutdown",
        )
        Runtime.getRuntime().addShutdownHook(shutdownHook)
        try {
            if (options.probeControlStdin) {
                awaitProbeShutdown()
                server.close()
                ProbeJson.write(
                    server.phaseThreeProbe().snapshot(),
                    server.phaseThreeProbe().maxConcurrentRuntimeOperations(),
                    System.out,
                )
            } else {
                server.awaitTermination()
            }
        } finally {
            try {
                Runtime.getRuntime().removeShutdownHook(shutdownHook)
            } catch (_: IllegalStateException) {
            }
            server.close()
        }
    }

    @Throws(Exception::class)
    private fun runBootstrap(options: Options, shaderpackRoot: Path) {
        val runtime = FakeRuntimeAdapter()
        try {
            VibrisBootstrap.start(
                VibrisBootstrap.Config(
                    options.port,
                    options.pendingRoot,
                    options.artifactRoot,
                    shaderpackRoot,
                ),
                runtime,
            ).use {
                awaitProbeShutdown()
                System.out.println("{\"type\":\"ShutdownComplete\",\"code\":\"OK\"}")
            }
        } catch (failure: VibrisBootstrap.Failure) {
            System.out.println(
                "{\"type\":\"StartupFailed\",\"code\":\"" + failure.code().name + "\"}",
            )
        }
    }

    @Throws(Exception::class)
    private fun awaitProbeShutdown() {
        val input = BufferedReader(InputStreamReader(System.`in`))
        while (true) {
            var line = input.readLine() ?: return
            if (line.isNotEmpty() && line[0] == '\uFEFF') {
                line = line.substring(1)
            }
            if (line == "shutdown") {
                return
            }
            if (line.isNotBlank()) {
                throw IllegalArgumentException("Unknown probe control command: $line")
            }
        }
    }

    private fun close(server: FakeVibrisServer) {
        try {
            server.close()
        } catch (_: InterruptedException) {
            Thread.currentThread().interrupt()
        }
    }

    @JvmRecord
    private data class Options(
        val port: Int,
        val workRoot: Path,
        val pendingRoot: Path,
        val artifactRoot: Path,
        val shaderpackRoot: Path?,
        val probeControlStdin: Boolean,
    ) {
        companion object {
            fun parse(arguments: Array<String>): Options {
                var port: Int? = null
                var workRoot: Path? = null
                var pendingRoot: Path? = null
                var artifactRoot: Path? = null
                var shaderpackRoot: Path? = null
                var probeControlStdin = false
                var index = 0
                while (index < arguments.size) {
                    when (arguments[index]) {
                        "--port" -> {
                            index++
                            port = requireValue(arguments, index, "--port").toInt()
                        }
                        "--work-root" -> {
                            index++
                            workRoot = Path.of(requireValue(arguments, index, "--work-root"))
                        }
                        "--pending-root" -> {
                            index++
                            pendingRoot = Path.of(requireValue(arguments, index, "--pending-root"))
                        }
                        "--artifact-root" -> {
                            index++
                            artifactRoot = Path.of(requireValue(arguments, index, "--artifact-root"))
                        }
                        "--shaderpack-root" -> {
                            index++
                            shaderpackRoot = Path.of(requireValue(arguments, index, "--shaderpack-root"))
                        }
                        "--probe-control-stdin" -> probeControlStdin = true
                        else -> throw IllegalArgumentException("Unknown argument: ${arguments[index]}")
                    }
                    index++
                }
                val actualPort = port
                    ?: throw IllegalArgumentException("Missing required argument: --port")
                val actualWorkRoot = workRoot
                    ?: throw IllegalArgumentException("Missing required argument: --work-root")
                val root = actualWorkRoot.toAbsolutePath().normalize()
                val resolvedPending = pendingRoot ?: root.resolve("pending-shaders")
                val resolvedArtifacts = artifactRoot ?: root.resolve("artifacts")
                if (shaderpackRoot != null && !probeControlStdin) {
                    throw IllegalArgumentException("--shaderpack-root requires --probe-control-stdin")
                }
                return Options(
                    actualPort,
                    root,
                    resolvedPending,
                    resolvedArtifacts,
                    shaderpackRoot,
                    probeControlStdin,
                )
            }

            private fun requireValue(
                arguments: Array<String>,
                index: Int,
                option: String,
            ): String {
                if (index >= arguments.size) {
                    throw IllegalArgumentException("Missing value for $option")
                }
                return arguments[index]
            }
        }
    }
}